### ISSUE TITLE
[$4000] OnxyData is being dropped from the notification payload

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14715]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open mweb chrome from User A account
2. Go to chat section with User B
3. Type in multi lines (6-7 lines) of emoji within a single message and send it to User B (Make sure to login with user B on android)
4. Notice that multilines of emojis sent from user A doesn’t show on notification on user B android

## Expected Result:
 Sending multi lines of emoji within a single message should be shown in the notification section like texts

## Actual Result:
Sending multi lines of emoji within a single message is not shown on the notification

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native

**Version Number:** 1.2.63-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/215894212-df5af399-2f2d-441d-8e90-81f36dacf9c2.mp4

https://user-images.githubusercontent.com/43996225/215894266-ad88c031-7c77-4fc0-90c5-f869c3fcfaa7.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675166101726659

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Tasks**
- https://github.com/Expensify/App/issues/18071 
- https://github.com/Expensify/Expensify/issues/286640
- https://github.com/Expensify/Expensify/issues/286641
- https://github.com/Expensify/Expensify/issues/286643

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0145bce78b355295bd</li>
        <li>Upwork Job ID: 1621105271543308288</li>
        <li>Last Price Increase: 2023-02-16</li>
    </ul>
</details>

**Comment by dylanexpensify** [comment]
reviewing

**Comment by priyeshshah11** [comment]
This could be due to some size limit in notifications' data.

**Comment by alexxxwork** [comment]
Expensify uses urbanairship lib to handle notifications
https://github.com/Expensify/App/blob/aea38c5589521101938dd988de67b0b8f5b23f54/src/libs/Notification/PushNotification/index.native.js#L59
It seems to truncate ordinary text to 112 ascii symbols like this:
⁦1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012..."
but can't truncate emoji's
if a line ends with other symobls - it will be displayed in notification
but if there's only emojis - you can recieve only 28 emojis in notification which is two lines of 18 symobls (may be device specific)

It seems to be the limits of urbanairship implementation as documented here:
https://support.airship.com/hc/en-us/articles/213491643-What-are-the-Maximum-Characters-for-Push-Notifications-

**Comment by dylanexpensify** [comment]
No proposals

**Comment by alexxxwork** [comment]
> No proposals

I think the problem is on the server side (airship push server).

**Comment by cead22** [comment]
Sorry I haven't looked into this but I'll prioritize it tomorrow

**Comment by alexxxwork** [comment]
@dylanexpensify updated my proposal with the template. Maybe this makes more sense now :)

**Comment by cead22** [comment]
I think we should confirm if this is in fact an airship notification. I think it's possible that we're exceeding 37-44 maximum character limit when sending more than 45 emojis, each of which has a string length > 2 because they're unicode characters

<img width="237" alt="image" src="https://user-images.githubusercontent.com/165133/217695487-50ea4c49-2308-419a-81d3-762cd82b242d.png">

Once we confirm that's the case by finding the limit for an android device, and confirming a message with 1 less character makes the notification work, then I think we can discuss whether it's worth truncating these messages ourselves, and how to do it

**Comment by cead22** [comment]
I wrote the previous comment assuming the column `notification drawer SystemNotification` was the right one, but is that the one to look at or `Notification drawer DefaultNotification Big Text Style`? `Lock Screen Portrait` has the same values as `notification drawer SystemNotification`

**Comment by alexxxwork** [comment]
@cead22 oh, thanks for this addition, I just didn't realized that there are more than 2-byte emojis. I'll make additional tests and post the results.

**Comment by cead22** [comment]
I thought I remembered there being some with length = 4, but I was very surprised by the one that returns 11

**Comment by alexxxwork** [comment]
> I thought I remembered there being some with length = 4, but I was very surprised by the one that returns 11

And there could be more when you change the skin color.
Here's 15-bytes: ``👩🏿‍❤️‍💋‍👨🏾`.length` (skin color - black)

**Comment by alexxxwork** [comment]
#### Updates on tests
So, posting test updates with new information about emojis
I'm making this test on a physical android device running android 13. Maybe this works different on a different version of android or physical screen/device.
Tested on production web.

1. ASCII single-string
You can send any string and get a notification. But if you have more than 110-115 symbols they get truncated like so:
`1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345..."`
Here's another bug - quote here is incosistent (have seen it on slack, I will search for the link)
Suprisingly certain amount of symbols in string that is truncated differs and I don't know why.
2. ASCII multi-string
Such messages are truncated also on 110-115 symbols counting added \n symbols. 
If you send short string of text (1\n2\n3\n4\n5\n6\n7\n8\n9\n0\n1\n2) 12 lines of text displays. And truncation doesn't work on the number of lines, additional text is just cut out. Also, if you have exactly 13 lines - they are shown in notification
Maybe this should be an additional bug
3. unicode characters single-string
JS uses UTF-16 so all unicode characters length === 1. But you can send only 38 utf-16 symbols like `茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶茶`. if the string is >38 symbols there's no notification (also seen a bug in slack)
If you add ascii symbols in the end of 38 unicode symbols - they are truncated on the 39th symbol like so `...茶茶茶a..."`
If you add ascii symbols in between and you have 38 unicode it won't display (seems like can't truncate)
4. unicode characters multi-string
Same as ascii - but if you send more then 13 lines - they won't show in notification
5. emojis 2-symbol (ordinary)
they behave like unicode but notification display only max 28 symbols as I wrote before
6. emojis 8-symbol
only 5-symbols. And those with 5 symbols could come with big delay like 1 minute.
8. emojis 15-symbol
only 3 symbols: `👩🏿‍❤️‍💋‍👨🏾👩🏿‍❤️‍💋‍👨🏾👩🏿‍❤️‍💋‍👨🏾`. You can add ascii symbols, they will be truncated. Managed to get this emojis split on truncation like on screen:
![Screenshot_20230211_002017_One UI Home](https://user-images.githubusercontent.com/3872759/218179316-5b7be7b9-e591-4a25-bd86-1ff42be4a351.jpg)

So, I think these weird bugs are server-side and the limit is in bytes around 110 bytes for the message. Server use optimal encoding so ascii is 1 byte, unicode 2 bytes. If we save `👩‍❤️‍👩` in a text file - it's 20 bytes. In hex this is: `f0 9f 91 a9 e2 80 8d e2 9d a4 ef b8 8f e2 80 8d f0 9f 91`. But JS thinks that  `👩‍❤️‍👩`.length === 8.

**Comment by dylanexpensify** [comment]
Not overdue - what are next steps here @eVoloshchak ?

**Comment by eVoloshchak** [comment]
@alexxxwork, wow, thank you for the thorough investigation!

> what are next steps here @eVoloshchak ?

@dylanexpensify, it seems like this is indeed a server-side [limitation](https://support.airship.com/hc/en-us/articles/213491643-What-are-the-Maximum-Characters-for-Push-Notifications-), external contributors can't work on this. 

I can confirm this happening on a physical Android device too, as well as the issue @alexxxwork was talking about [here](https://github.com/Expensify/App/issues/14715#issuecomment-1426256907):
> if the string is >38 symbols there's no notification (also seen a bug in slack)

**Comment by MelvinBot** [comment]
@cead22 @eVoloshchak @dylanexpensify this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by alexxxwork** [comment]
@eVoloshchak I've made a proposal in similar issue to truncate notification strings to Airship on our side here: https://github.com/Expensify/App/issues/14849 
I this a valid solution to this issue or should we address it upstream to Airship service?

**Comment by cead22** [comment]
Thanks for the detailed testing, there's a couple of things I didn't quite understand, but I'm asking Airship about any limitations or functionality they may have for truncating messages. I'll keep this issue updated with what I get

**Comment by cead22** [comment]
> Hi @carlos I asked internally and was told that we do not have any functionality to truncate messages. But also, it's not expected the device would not receive the notification. The team recommended enabling verbose logging to see if the SDK is receiving a push or not.  It would also help if you could provide some examples of messages that aren’t coming through. We may want to open a ticket with support to take a look into it.

I shared the test video in this issue with them, but can you see if you can enable verbose logging (if not already enabled for dev) and check if there are any logs about the push notification in question?

**Comment by dylanexpensify** [comment]
@cead22 shall we move this internal?

**Comment by cead22** [comment]
I think so, at least for now until we're sure whether this can be fixed externally

**Comment by MelvinBot** [comment]
Current assignee @eVoloshchak is eligible for the Internal assigner, not assigning anyone new.

**Comment by alexxxwork** [comment]
> I shared the test video in this issue with them, but can you see if you can enable verbose logging (if not already enabled for dev) and check if there are any logs about the push notification in question?

Sorry for the delay. Would it help to post metro logs? There are just nothing on missing notifications for the cases I mentioned in my investigation https://github.com/Expensify/App/issues/14715#issuecomment-1414433837 (multi-lined unicode and emojis strings)

**Comment by cead22** [comment]
They asked if we could provide logs from the Airship SDK and linked to this documentation page https://docs.airship.com/platform/mobile/logging/?groupid=ios-swift. Is that something you can pull?

**Comment by alexxxwork** [comment]
> They asked if we could provide logs from the Airship SDK and linked to this documentation page https://docs.airship.com/platform/mobile/logging/?groupid=ios-swift. Is that something you can pull?

@cead22 shure. Logs attached. [2023-02-22 00:31:38.293 17245-17245 New Expensify.log](https://github.com/Expensify/App/files/10797266/2023-02-22.00.31.38.293.17245-17245.New.Expensify.log)
<details>
<summary>Screens for this dialogue</summary>


</details>

**Comment by MelvinBot** [comment]
@cead22 @eVoloshchak @dylanexpensify this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a>. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by cead22** [comment]
Thanks for the logs, I sent them to airship, and I'm still waiting on a response from them

**Comment by cead22** [comment]
They got back to me with the following -- which I think is wrong, but please check my logic (my reply is below)

>@carlos It looks like the device is receiving the push and we are trying to post it, assuming its this push:
2023-02-22 01:06:52.288 17245-22160 New Expensify - UALib   com.expensify.chat                   I  Processing push: {fromExpensify=true, com.urbanairship.metadata=eyJ2ZXJzaW9uX2lkIjoxLCJ0aW1lIjoxNjc3MDEwMDExOTIzLCJwdXNoX2lkIjoiNDY2ZTgxMDAtYjIyMy0xMWVkLTk3NTUtMDI0MjY0ZWY3ZDE5IiwiY2FtcGFpZ25zIjp7ImNhdGVnb3JpZXMiOltdfX0=, com.urbanairship.push.APID=0bd51415-27ec-48c4-a8cc-7767b80281d7, payload={"title":"spam@vcc.vc","type":"reportComment","reportID":4148003058225040,"shouldScrollToLastUnread":true,"roomName":null,"onyxData":[{"onyxMethod":"merge","key":"report_4148003058225040","value":{"reportID":"4148003058225040","maxSequenceNumber":47,"lastActionCreated":"2023-02-21 20:06:48.754","lastVisibleActionCreated":"2023-02-21 20:06:48.754","lastReadTime":"2023-02-21 20:01:24.544","lastMessageText":"\u20660123456789012345678901234567890123456789012345678901234567890123456789012345678","lastActorEmail":"spam@vcc.vc"}},{"onyxMethod":"merge","key":"reportActions_4148003058225040","value":{"2105426769474541683":{"person":[{"type":"TEXT","style":"strong","text":"spam@vcc.vc"}],"actorEmail":"spam@vcc.vc","actorAccountID":13031568,"message":[{"type":"COMMENT","html":"\u206601234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789","text":"\u206601234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789","isEdited":false}],"originalMessage":{"html":"\u206601234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_4.png","created":"2023-02-21 20:06:48.754","timestamp":1677010008,"reportActionTimestamp":1677010008754,"automatic":false,"sequenceNumber":47,"actionName":"ADDCOMMENT","shouldShow":true,"clientID":"","reportActionID":"2105426769474541683","isAttachment":false}},"shouldNotify":true}]}, com.urbanairship.push.CANONICAL_PUSH_ID=466e8100-b223-11ed-9755-024264ef7d19, com.urbanairship.push.PUSH_ID=47f5bd40-b223-11ed-8d98-02421fc6a487, com.urbanairship.notification_channel=4148003058225040, com.urbanairship.push.ALERT=⁦0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901...", com.urbanairship.title=spam@vcc.vc}
That means its probably an issue with either this - https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java  but it looks fine to me or an OS issue. Are you able to reproduce on other devices?

[CustomNotificationProvider.java](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java)

```
package com.expensify.chat.customairshipextender;
import static androidx.core.app.NotificationCompat.CATEGORY_MESSAGE;
import static androidx.core.app.NotificationCompat.PRIORITY_MAX;
import android.app.NotificationChannel;
import android.app.NotificationChannelGroup;
import android.app.NotificationManager;
```
>You might be able to see a helpful error message in the notification service if you remove the filter on the app


---
### My Reply

```
I don't think that's the right push, I think it's this one (I noticed Notification dismissed in there), since this is the one with a ton of emojis, and the one with a ton of emojis is the one we're not seeing. I'll check about the devices, I know this is happening on android but I'm not sure which specific devices
459 2023-02-22 01:02:17.381 17245-17245 New Expensify - UALib   com.expensify.chat                   V  NotificationIntentProcessor - Processing intent: com.urbanairship.push.ACTION_NOTIFICATION_DISMISSED
460 2023-02-22 01:02:17.381 17245-17245 New Expensify - UALib   com.expensify.chat                   I  Notification dismissed: NotificationInfo{alert=:grinning::grinning::grinning::grinning:
461                                                                                                     :grinning::grinning::grinning::grinning:
462                                                                                                     :grinning::grinning::grinning::grinning:
463                                                                                                     :grinning::grinning::grinning::grinning:
464                                                                                                     :grinning::grinning::grinning::grinning:
465                                                                                                     :grinning::grinning::grinning::grinning:, notificationId=1007, notificationTag='null'}
466 2023-02-22 01:02:17.382 17245-20975 New Expensify - UALib   com.expensify.chat                   V  NotificationProxyReceiver - Finished processing notification intent with result true.
```

>Looks like this in the file actually, but I think slack is changing the emojis to text
![image](https://user-images.githubusercontent.com/165133/221064226-18a169b9-487d-407c-8310-51fcb1f7a23d.png)

**Comment by alexxxwork** [comment]
@cead22 they got wrong log line. This is a truncated push with numbers, and we are talking about 13 lines of emojis.

Also, I made a service request to them in another issue and it seems that they just mixing the problems.

**Comment by alexxxwork** [comment]
@cead22 in this issue https://github.com/Expensify/App/issues/14849 we are dealing with similar but different problem - Airship truncates strings and adds additional `...''` in the end of the line. So @parasharrajat asked to make a service request to Airship and I made one https://support.airship.com/hc/en-us/requests/89947 where I have provided the same logs and details about https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java . So it seems that they are trying to merge our requests to one issue, but these are different cases :)

**Comment by dylanexpensify** [comment]
@cead22 let me know if there are any next steps here I can action!

**Comment by fabioh8010** [comment]
Hi, I'm Fábio from Callstack and I would like to take a look at this issue.

**Comment by cead22** [comment]
@alexxxwork thanks for your help so far. We're going to take this internal and work with Callstack and Airship to resolve this given that it has gone 4+ weeks, but we'll be compensating you for your help so far

**Comment by MelvinBot** [comment]
@cead22 @eVoloshchak @dylanexpensify this issue is now 4 weeks old and preventing us from maintaining WAQ. This should now be your highest priority. Please post below what your plan is to get a PR in review ASAP. Thanks!

**Comment by cead22** [comment]
I'm connecting Callstack and Airship on slack to help push this forward

**Comment by fabioh8010** [comment]
I'll be posting here the investigations I did and that were also posted on Airship Slack's channel.

---

So, I did some investigation focusing exclusively on native side to debug and understand why we are having this issues on Android.

I did four tests basically:
1. Send a short message
2. Send a long message
3. Send a short message with emojis only
4. Send a long message with emojis only

<details>
<summary><b>Test 1 - Send a short message</b></summary>

I sent this message to the account that I’m using in my Android emulator.

```
Test
```

![Screenshot 2023-02-28 at 14 11 38](https://user-images.githubusercontent.com/20051562/222470605-a7d9ef05-6460-4665-86d8-117d5afd2676.png)

As you can see in my screenshot, the notification was sent to my emulator and it’s correctly styled.

We implemented a [custom provider](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java) in our app to handle the notifications, and at [this line](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java#L83) he’ll have the following payload:

```json
{
  "title": "fabio.henriques+1@callstack.com",
  "type": "reportComment",
  "reportID": 4530354462045455,
  "reportAction": {
    "person": [
      {
        "type": "TEXT",
        "style": "strong",
        "text": "fabio.henriques+1@callstack.com"
      }
    ],
    "actorEmail": "fabio.henriques+1@callstack.com",
    "actorAccountID": 14356954,
    "message": [
      {
        "type": "COMMENT",
        "html": "Test",
        "text": "Test",
        "isEdited": false
      }
    ],
    "originalMessage": {
      "html": "Test"
    },
    "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_3.png",
    "created": "2023-02-28 13:45:24.733",
    "timestamp": 1677591924,
    "reportActionTimestamp": 1677591924733,
    "automatic": false,
    "sequenceNumber": 59,
    "actionName": "ADDCOMMENT",
    "shouldShow": true,
    "reportActionID": "6374191708963222803"
  },
  "shouldScrollToLastUnread": true,
  "roomName": null,
  "onyxData": [
    {
      "onyxMethod": "merge",
      "key": "report_4530354462045455",
      "value": {
        "reportID": "4530354462045455",
        "maxSequenceNumber": 59,
        "lastVisibleActionCreated": "2023-02-28 13:45:24.733",
        "lastReadTime": "2023-02-28 09:57:09.605",
        "lastMessageText": "Test",
        "lastActorEmail": "fabio.henriques+1@callstack.com"
      }
    },
    {
      "onyxMethod": "merge",
      "key": "reportActions_4530354462045455",
      "value": {
        "6374191708963222803": {
          "person": [
            {
              "type": "TEXT",
              "style": "strong",
              "text": "fabio.henriques+1@callstack.com"
            }
          ],
          "actorEmail": "fabio.henriques+1@callstack.com",
          "actorAccountID": 14356954,
          "message": [
            {
              "type": "COMMENT",
              "html": "Test",
              "text": "Test",
              "isEdited": false
            }
          ],
          "originalMessage": {
            "html": "Test"
          },
          "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_3.png",
          "created": "2023-02-28 13:45:24.733",
          "timestamp": 1677591924,
          "reportActionTimestamp": 1677591924733,
          "automatic": false,
          "sequenceNumber": 59,
          "actionName": "ADDCOMMENT",
          "shouldShow": true,
          "reportActionID": "6374191708963222803",
          "isAttachment": false
        }
      },
      "shouldNotify": true
    }
  ]
}
```

![Screenshot 2023-02-28 at 14 43 18](https://user-images.githubusercontent.com/20051562/222470924-bcf93a28-8dd5-4f38-8c9b-ff02f0285c66.png)

We can also see in the device logs that the library is processing the notification:

```
New Expensify - UALib: Processing push: {fromExpensify=true, com.urbanairship.metadata=eyJ2ZXJzaW9uX2lkIjoxLCJ0aW1lIjoxNjc3NTkxOTMxMDY2LCJwdXNoX2lkIjoiMjc3NjNmZTAtYjc2ZS0xMWVkLTlmODAtMDI0MjhhNzllZDMxIiwiY2FtcGFpZ25zIjp7ImNhdGVnb3JpZXMiOltdfX0=, com.urbanairship.push.APID=29a18e78-47f7-4fff-b063-9f4fe54c9b4d, payload={"title":"fabio.henriques+1@callstack.com","type":"reportComment","reportID":4530354462045455,"reportAction":{"person":[{"type":"TEXT","style":"strong","text":"fabio.henriques+1@callstack.com"}],"actorEmail":"fabio.henriques+1@callstack.com","actorAccountID":14356954,"message":[{"type":"COMMENT","html":"Test","text":"Test","isEdited":false}],"originalMessage":{"html":"Test"},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_3.png","created":"2023-02-28 13:45:24.733","timestamp":1677591924,"reportActionTimestamp":1677591924733,"automatic":false,"sequenceNumber":59,"actionName":"ADDCOMMENT","shouldShow":true,"reportActionID":"6374191708963222803"},"shouldScrollToLastUnread":true,"roomName":null,"onyxData":[{"onyxMethod":"merge","key":"report_4530354462045455","value":{"reportID":"4530354462045455","maxSequenceNumber":59,"lastVisibleActionCreated":"2023-02-28 13:45:24.733","lastReadTime":"2023-02-28 09:57:09.605","lastMessageText":"Test","lastActorEmail":"fabio.henriques+1@callstack.com"}},{"onyxMethod":"merge","key":"reportActions_4530354462045455","value":{"6374191708963222803":{"person":[{"type":"TEXT","style":"strong","text":"fabio.henriques+1@callstack.com"}],"actorEmail":"fabio.henriques+1@callstack.com","actorAccountID":14356954,"message":[{"type":"COMMENT","html":"Test","text":"Test","isEdited":false}],"originalMessage":{"html":"Test"},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_3.png","created":"2023-02-28 13:45:24.733","timestamp":1677591924,"reportActionTimestamp":1677591924733,"automatic":false,"sequenceNumber":59,"actionName":"ADDCOMMENT","shouldShow":true,"reportActionID":"6374191708963222803","isAttachment":false}},"shouldNotify":true}]}, com.urbanairship.push.CANONICAL_PUSH_ID=27763fe0-b76e-11ed-9f80-02428a79ed31, com.urbanairship.push.PUSH_ID=2acc21a0-b76e-11ed-aef1-0242e83d77e5, com.urbanairship.notification_channel=4530354462045455, com.urbanairship.push.ALERT=Test, com.urbanairship.title=fabio.henriques+1@callstack.com}
```

So far so good, the payload is fine, it will be passed to our applyMessageStyle() method [here](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java#L101) in order to [process it](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java#L159) adding the avatar image and add this “Chat with...” text to the title, among with other things.

And it’s done, the notification will be shown as we expect.

</details>

<details>
<summary><b>Test 2 - Send a long message</b></summary>

I sent this message to the account that I’m using in my Android emulator.

```
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

![Screenshot 2023-02-28 at 14 12 13](https://user-images.githubusercontent.com/20051562/222471834-ad58d9e0-b8ed-41cd-be09-b8fbb8392824.png)

In my screenshot, the notification sent to my device has the problem discussed in [this issue](https://github.com/Expensify/App/issues/14849): the text is truncated and with a `"` at the end, and also we can notice that the notification is not styled as the first one with the short message, it’s missing the avatar image and the title with “Chat with...” text.

Also, for some reason, the payload is very small this time, missing two important properties (`reportAction` and `onyxData`) that we’ll need in our custom provider later:

```json
{
  "title": "fabio.henriques+1@callstack.com",
  "type": "reportComment",
  "reportID": 4530354462045455,
  "shouldScrollToLastUnread": true,
  "roomName": null
}
```

![Screenshot 2023-02-28 at 14 44 18](https://user-images.githubusercontent.com/20051562/222471873-2e9c8e15-a631-4bd0-bfaf-1173ef648efd.png)

Log:

```
New Expensify - UALib: Processing push: {fromExpensify=true, com.urbanairship.metadata=eyJ2ZXJzaW9uX2lkIjoxLCJ0aW1lIjoxNjc3NTkyMzcwMzY2LCJwdXNoX2lkIjoiMmVlOTA3NzAtYjc2Zi0xMWVkLWE1NTEtMDI0Mjc1MWZiZjQ5IiwiY2FtcGFpZ25zIjp7ImNhdGVnb3JpZXMiOltdfX0=, com.urbanairship.push.APID=29a18e78-47f7-4fff-b063-9f4fe54c9b4d, payload={"title":"fabio.henriques+1@callstack.com","type":"reportComment","reportID":4530354462045455,"shouldScrollToLastUnread":true,"roomName":null}, com.urbanairship.push.CANONICAL_PUSH_ID=2ee90770-b76f-11ed-a551-0242751fbf49, com.urbanairship.push.PUSH_ID=30a3fde0-b76f-11ed-858e-02423fc6104c, com.urbanairship.notification_channel=4530354462045455, com.urbanairship.push.ALERT=Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna...", com.urbanairship.title=fabio.henriques+1@callstack.com}
```

Also please notice that this `com.urbanairship.push.ALERT` property from the log has this value `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna..."`, precisely the truncated text with the quote that we have in the notification.

Because of that, a exception is thrown at [this line](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java#L166) and we end up not applying any custom stylings and leaving to the library to apply the default ones. Then, I presume the library will use the `com.urbanairship.push.ALERT` property as the body of the notification.

```
NotificationProvider: Failed to parse conversation. SendID=30a3fde0-b76f-11ed-858e-02423fc6104c
NotificationProvider: java.lang.NullPointerException: Attempt to invoke virtual method 'com.urbanairship.json.JsonMap com.urbanairship.json.JsonValue.getMap()' on a null object reference
NotificationProvider: 	at com.expensify.chat.customairshipextender.CustomNotificationProvider.applyMessageStyle(CustomNotificationProvider.java:166)
NotificationProvider: 	at com.expensify.chat.customairshipextender.CustomNotificationProvider.onExtendBuilder(CustomNotificationProvider.java:101)
NotificationProvider: 	at com.urbanairship.push.notifications.AirshipNotificationProvider.onCreateNotification(AirshipNotificationProvider.java:210)
NotificationProvider: 	at com.urbanairship.push.IncomingPushRunnable.postProcessPush(IncomingPushRunnable.java:196)
NotificationProvider: 	at com.urbanairship.push.IncomingPushRunnable.processPush(IncomingPushRunnable.java:144)
NotificationProvider: 	at com.urbanairship.push.IncomingPushRunnable.run(IncomingPushRunnable.java:97)
NotificationProvider: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:463)
NotificationProvider: 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
NotificationProvider: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
NotificationProvider: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
NotificationProvider: 	at com.urbanairship.util.AirshipThreadFactory$1.run(AirshipThreadFactory.java:50)
NotificationProvider: 	at java.lang.Thread.run(Thread.java:1012)
```

</details>

<details>
<summary><b>Test 3 - Send a short message with emojis only</b></summary>

I sent this message to the account that I’m using in my Android emulator.

```
🙂🙂🙂
```

![Screenshot 2023-02-28 at 14 03 56](https://user-images.githubusercontent.com/20051562/222472911-70d520f8-d7d1-4e0e-8e37-b6f3dc5a6b1f.png)

Like the first test, the notification will be shown correctly with our custom stylings, so no problem here.

```json
{
  "title": "fabio.henriques+1@callstack.com",
  "type": "reportComment",
  "reportID": 4530354462045455,
  "reportAction": {
    "person": [
      {
        "type": "TEXT",
        "style": "strong",
        "text": "fabio.henriques+1@callstack.com"
      }
    ],
    "actorEmail": "fabio.henriques+1@callstack.com",
    "actorAccountID": 14356954,
    "message": [
      {
        "type": "COMMENT",
        "html": ":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:",
        "text": ":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:",
        "isEdited": false
      }
    ],
    "originalMessage": {
      "html": ":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:"
    },
    "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_3.png",
    "created": "2023-02-28 14:03:17.657",
    "timestamp": 1677592997,
    "reportActionTimestamp": 1677592997657,
    "automatic": false,
    "sequenceNumber": 61,
    "actionName": "ADDCOMMENT",
    "shouldShow": true,
    "reportActionID": "6932692835455476580"
  },
  "shouldScrollToLastUnread": true,
  "roomName": null,
  "onyxData": [
    {
      "onyxMethod": "merge",
      "key": "report_4530354462045455",
      "value": {
        "reportID": "4530354462045455",
        "maxSequenceNumber": 61,
        "lastVisibleActionCreated": "2023-02-28 14:03:17.657",
        "lastReadTime": "2023-02-28 09:57:09.605",
        "lastMessageText": ":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:",
        "lastActorEmail": "fabio.henriques+1@callstack.com"
      }
    },
    {
      "onyxMethod": "merge",
      "key": "reportActions_4530354462045455",
      "value": {
        "6932692835455476580": {
          "person": [
            {
              "type": "TEXT",
              "style": "strong",
              "text": "fabio.henriques+1@callstack.com"
            }
          ],
          "actorEmail": "fabio.henriques+1@callstack.com",
          "actorAccountID": 14356954,
          "message": [
            {
              "type": "COMMENT",
              "html": ":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:",
              "text": ":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:",
              "isEdited": false
            }
          ],
          "originalMessage": {
            "html": ":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:"
          },
          "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_3.png",
          "created": "2023-02-28 14:03:17.657",
          "timestamp": 1677592997,
          "reportActionTimestamp": 1677592997657,
          "automatic": false,
          "sequenceNumber": 61,
          "actionName": "ADDCOMMENT",
          "shouldShow": true,
          "reportActionID": "6932692835455476580",
          "isAttachment": false
        }
      },
      "shouldNotify": true
    }
  ]
}
```

```
New Expensify - UALib: Processing push: {fromExpensify=true, com.urbanairship.metadata=eyJ2ZXJzaW9uX2lkIjoxLCJ0aW1lIjoxNjc3NTkzMDA0NjgxLCJwdXNoX2lkIjoiYTZmNjY4NjAtYjc3MC0xMWVkLTgzZWQtMDI0MmU1Y2FkN2Q0IiwiY2FtcGFpZ25zIjp7ImNhdGVnb3JpZXMiOltdfX0=, com.urbanairship.push.APID=29a18e78-47f7-4fff-b063-9f4fe54c9b4d, payload={"title":"fabio.henriques+1@callstack.com","type":"reportComment","reportID":4530354462045455,"reportAction":{"person":[{"type":"TEXT","style":"strong","text":"fabio.henriques+1@callstack.com"}],"actorEmail":"fabio.henriques+1@callstack.com","actorAccountID":14356954,"message":[{"type":"COMMENT","html":"\ud83d\ude42\ud83d\ude42\ud83d\ude42","text":"\ud83d\ude42\ud83d\ude42\ud83d\ude42","isEdited":false}],"originalMessage":{"html":"\ud83d\ude42\ud83d\ude42\ud83d\ude42"},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_3.png","created":"2023-02-28 14:03:17.657","timestamp":1677592997,"reportActionTimestamp":1677592997657,"automatic":false,"sequenceNumber":61,"actionName":"ADDCOMMENT","shouldShow":true,"reportActionID":"6932692835455476580"},"shouldScrollToLastUnread":true,"roomName":null,"onyxData":[{"onyxMethod":"merge","key":"report_4530354462045455","value":{"reportID":"4530354462045455","maxSequenceNumber":61,"lastVisibleActionCreated":"2023-02-28 14:03:17.657","lastReadTime":"2023-02-28 09:57:09.605","lastMessageText":"\ud83d\ude42\ud83d\ude42\ud83d\ude42","lastActorEmail":"fabio.henriques+1@callstack.com"}},{"onyxMethod":"merge","key":"reportActions_4530354462045455","value":{"6932692835455476580":{"person":[{"type":"TEXT","style":"strong","text":"fabio.henriques+1@callstack.com"}],"actorEmail":"fabio.henriques+1@callstack.com","actorAccountID":14356954,"message":[{"type":"COMMENT","html":"\ud83d\ude42\ud83d\ude42\ud83d\ude42","text":"\ud83d\ude42\ud83d\ude42\ud83d\ude42","isEdited":false}],"originalMessage":{"html":"\ud83d\ude42\ud83d\ude42\ud83d\ude42"},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_3.png","created":"2023-02-28 14:03:17.657","timestamp":1677592997,"reportActionTimestamp":1677592997657,"automatic":false,"sequenceNumber":61,"actionName":"ADDCOMMENT","shouldShow":true,"reportActionID":"6932692835455476580","isAttachment":false}},"shouldNotify":true}]}, com.urbanairship.push.CANONICAL_PUSH_ID=a6f66860-b770-11ed-83ed-0242e5cad7d4, com.urbanairship.push.PUSH_ID=aab8c790-b770-11ed-86bc-0242915033f4, com.urbanairship.notification_channel=4530354462045455, com.urbanairship.push.ALERT=:slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:, com.urbanairship.title=fabio.henriques+1@callstack.com}
```

</details>

<details>
<summary><b>Test 4 - Send a long message with emojis only</b></summary>

I sent this message to the account that I’m using in my Android emulator.

```
🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂🙂
```

This is the worst case, because in this test we didn’t receive any logs from Airship at all. It’s like it never hit the device. I can show in a video.

https://user-images.githubusercontent.com/20051562/222475428-063718c6-4228-4e14-8efb-cddaeb0cfa5a.mov

The log we got was from `Pusher`, but this is JS logic that only makes effect in web / desktop, it’s not related to Airship’s push notifications.

```
02-28 21:41:53.994 24278 24422 D ReactNativeJS: [info] [Report] Handled onyxApiUpdate event sent by Pusher - [{"onyxMethod":"merge","key":"report_4530354462045455","value":{"reportID":"4530354462045455","maxSequenceNumber":81,"lastVisibleActionCreated":"2023-02-28 21:41:54.465","lastReadTime":"2023-02-28 20:45:33.064","lastMessageText":":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:","lastActorEmail":"fabio.henriques+1@callstack.com"}},{"onyxMethod":"merge","key":"reportActions_4530354462045455","value":{"4605440722137264120":{"person":[{"type":"TEXT","style":"strong","text":"fabio.henriques+1@callstack.com"}],"actorEmail":"fabio.henriques+1@callstack.com","actorAccountID":14356954,"message":[{"type":"COMMENT","html":":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:","text":":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:","isEdited":false}],"originalMessage":{"html":":slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face::slightly_smiling_face:"},"avatar":"https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_3.png","created":"2023-02-28 21:41:54.465","timestamp":1677620514,"reportActionTimestamp":1677620514465,"automatic":false,"sequenceNumber":81,"actionName":"ADDCOMMENT","shouldShow":true,"reportActionID":"4605440722137264120","isAttachment":false}},"shouldNotify":true}]
02-28 21:41:53.995 24278 24422 D ReactNativeJS: [info] [LOCAL_NOTIFICATION] Creating notification - ""
```

</details>

---

I also tried to disable our custom provider and listener [here](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomAirshipExtender.java)

```java
public class CustomAirshipExtender implements AirshipExtender {
    @Override
    public void onAirshipReady(@NonNull Context context, @NonNull UAirship airship) {
        PushManager pushManager = airship.getPushManager();

        CustomNotificationProvider notificationProvider = new CustomNotificationProvider(context, airship.getAirshipConfigOptions());
        //pushManager.setNotificationProvider(notificationProvider);

        NotificationListener notificationListener = airship.getPushManager().getNotificationListener();
        // pushManager.setNotificationListener(new CustomNotificationListener(notificationListener, notificationProvider));
    }
}
```

But it doesn’t make difference about the problems, we are still going to get the truncated text and don’t receive the notification with a lot of emojis.

From my understanding, the problem is not in the library or our custom provider, it’s **before** the notification hit the device (or not hit, like the Test 4). Even debugging some library’s methods, it all seems fine to me.

---

Airship's team told us that it could be a issue with the size of our payloads. FCM payloads have a 4000 byte limit, and the payload from **Test 1** has almost 2k bytes. I took this payload and replaced every `"Test"` occurence with `"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."` in order to reproduce what could be the payload from **Test 2**, and this payload ended up with more than 4k bytes, which could explain why the large message (**Test 2**) didn’t work properly.

**Comment by alexxxwork** [comment]
> Airship's team told us that it could be a issue with the size of our payloads. FCM payloads have a 4000 byte limit

@fabioh8010 I've tested long plain-text payloads and they always worked (comes truncated). The problem is in symbols that are encoded as unicode (not only emojis) see my report https://github.com/Expensify/App/issues/14715#issuecomment-1426256907 . I've counted the maximum symbol amount that works correctly: 38 UTF-16 symbols. If you add some plain ascii symbols in the end - the message would be truncated and would come as push message.

**Comment by fabioh8010** [comment]
@alexxxwork The problem is not only with symbols. We are indeed dealing with big payloads being sent with the notification. For example, this is the payload of a simple `Test` message.

```json
{
  "title": "fabio.henriques+1@callstack.com",
  "type": "reportComment",
  "reportID": 4530354462045455,
  "reportAction": {
    "person": [
      {
        "type": "TEXT",
        "style": "strong",
        "text": "fabio.henriques+1@callstack.com"
      }
    ],
    "actorEmail": "fabio.henriques+1@callstack.com",
    "actorAccountID": 14356954,
    "message": [
      {
        "type": "COMMENT",
        "html": "Test",
        "text": "Test",
        "isEdited": false,
        "reactions": []
      }
    ],
    "originalMessage": {
      "html": "Test"
    },
    "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_3.png",
    "created": "2023-03-03 18:29:28.126",
    "timestamp": 1677868168,
    "reportActionTimestamp": 1677868168126,
    "automatic": false,
    "sequenceNumber": 104,
    "actionName": "ADDCOMMENT",
    "shouldShow": true,
    "reportActionID": "407009459185780717"
  },
  "shouldScrollToLastUnread": true,
  "roomName": null,
  "onyxData": [
    {
      "onyxMethod": "merge",
      "key": "report_4530354462045455",
      "value": {
        "reportID": "4530354462045455",
        "maxSequenceNumber": 104,
        "lastVisibleActionCreated": "2023-03-03 18:29:28.126",
        "lastReadTime": "2023-03-03 18:29:20.345",
        "lastMessageText": "Test",
        "lastActorEmail": "fabio.henriques+1@callstack.com"
      }
    },
    {
      "onyxMethod": "merge",
      "key": "reportActions_4530354462045455",
      "value": {
        "407009459185780717": {
          "person": [
            {
              "type": "TEXT",
              "style": "strong",
              "text": "fabio.henriques+1@callstack.com"
            }
          ],
          "actorEmail": "fabio.henriques+1@callstack.com",
          "actorAccountID": 14356954,
          "message": [
            {
              "type": "COMMENT",
              "html": "Test",
              "text": "Test",
              "isEdited": false,
              "reactions": []
            }
          ],
          "originalMessage": {
            "html": "Test"
          },
          "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_3.png",
          "created": "2023-03-03 18:29:28.126",
          "timestamp": 1677868168,
          "reportActionTimestamp": 1677868168126,
          "automatic": false,
          "sequenceNumber": 104,
          "actionName": "ADDCOMMENT",
          "shouldShow": true,
          "reportActionID": "407009459185780717",
          "isAttachment": false
        }
      },
      "shouldNotify": true
    }
  ]
}
```

This one has almost 2k bytes for just a simple message. The FCM limit for payloads is 4k and Airship's team stated that they also add some internal properties that may increase the total payload size. There are a lot of properties that are not being used from the App side and could be stripped from the payload. These are the current ones being used:

**JS side**
- reportID
- onyxData

**Android custom provider (styling)**
- reportID
- reportAction.person[0].text
- reportAction.avatar
- reportAction.actorAccountID
- reportAction.message[0].text
- reportAction.created
- roomName

Also, they suggested us to use gzip compression to achieve smaller payloads if necessary.

But I agree with you that we are facing other problems as well like the ones you showed in your investigation. To summarize:
- Big payloads are being sent with the notifications even for simple messages. We could reduce these payloads by only sending the necessary data and also applying gzip compression if necessary.
- When the message is larger the payload comes malformed, sometimes missing the `reportAction` and/or `onyxData` properties, also the message is already coming truncated and with the `"` at the end. Because of the malformed payload a exception is thrown in our Android custom notification provider and our custom stylings end up not being applied to the notification.
- Sometimes the notifications are not sent at all:
  - If you send a large message with only emojis.
  - The situations described by @alexxxwork investigations.

From my perspective, we are dealing with two issues:
1. Big payloads that could be reduced and/or compressed
2. Encoding or other issues in the backend that can be making the notification be sent malformed or not sent at all

cc @cead22

**Comment by alexxxwork** [comment]
@fabioh8010 You are describing payload being sent to Expensify API. I think someone from Expensify backend team should confirm that the same payload is being sent to Airship. (@tgolen ?) 
You can try and send up to 15000 symbols in plaintext in a message and it would work, the truncated push would come, so the problem is in the Airship truncation mechanics.

**Comment by fabioh8010** [comment]
Hi @alexxxwork, these payloads are what the mobile application receives when we receive a notification. We use the information from these payloads to customise Android's notification [here](https://github.com/Expensify/App/blob/main/android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java#L101) and to handle its data when it's received or tapped [here](https://github.com/Expensify/App/blob/main/src/libs/actions/Report.js#L56)

**Comment by dylanexpensify** [comment]
Not overdue, we're working on it

**Comment by Julesssss** [comment]
I started by looking through our backend notification logic. First of all, the quote message issue is caused by internal logic that I'm not sure needs to apply to NewDot. I'm looking into this first as it will help me to resolve [this related internal issue ](https://github.com/Expensify/Expensify/issues/260138)(Android notification grouping).

**Comment by Julesssss** [comment]
Nuked Android Studio, problem solved... 

I've been debugging both the backend and Android and we have _some_ options for reducing payload data. First I'll roll back the initial limit we had previously set to fix the quote issue, then I'll work on further reducing the payload.

**Comment by alexxxwork** [comment]
> First of all, the quote message issue is caused by internal logic that I'm not sure needs to apply to NewDot.

@Julesssss Could you confirm that `...''` ending is coming from E/App backend and has nothing to do with Airship? I thnik I should close the issue, opened in Airship about this bug.

**Comment by Julesssss** [comment]
@alexxxwork yeah the ellipsised message is 100% our internal issue. Please go ahead and close the airship bug.

**Comment by Julesssss** [comment]
I'm making good progress here, but the changes are slow to test as I need debuggable release builds which are slow to compile. [I'm working with Ryan from Airship](https://expensify.slack.com/archives/C01RJ1RCK8D/p1678462368107329?thread_ts=1676655235.172589&cid=C01RJ1RCK8D) to help increase our payload allowance in a few ways:

- Increasing our 2kb limit to 3kb+
- Removing payload data that is not necessary for NewDot
- Ignoring some of the Airship fields that we can reconstruct natively from the onyxData 

We'll need to deploy the APP PR first, to ensure as many users are migrated to the new onyx data source before we deprecate fields that the app currently relies on

**Comment by Julesssss** [comment]
Ryan from Airship has gone one step further and is[ going to provide a way to gzip](https://expensify.slack.com/archives/C01RJ1RCK8D/p1678470995448979?thread_ts=1676655235.172589&cid=C01RJ1RCK8D) our payload, reducing the size by a decent amount!

**Comment by dylanexpensify** [comment]
Amazing!

**Comment by Julesssss** [comment]
[PR in review](https://github.com/Expensify/App/pull/15857)

**Comment by Julesssss** [comment]
Actually, the PR is just one of a few PRs necessary to completely resolve this one

**Comment by Julesssss** [comment]
IN PROGRESS

**Comment by Julesssss** [comment]
The first [PR](https://github.com/Expensify/App/pull/15857) has just hit prod, more changes are incoming.

**Comment by dylanexpensify** [comment]
Nice, not overdue, still moving

**Comment by MelvinBot** [comment]
@Julesssss, @dylanexpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@Julesssss, @dylanexpensify Huh... This is 4 days overdue. Who can take care of this?

**Comment by Julesssss** [comment]
No update, the notification issue is still outstanding.

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @Julesssss, @dylanexpensify eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by Julesssss** [comment]
Held because Manual Requests is super critical for EC3 and we're at risk of not being ready. Will be returning to this after.

**Comment by dylanexpensify** [comment]
@Julesssss shall we get this back to daily now?

**Comment by Julesssss** [comment]
Sorry I'm still insanely busy with Manual requests bugs. Aiming for next week

**Comment by dylanexpensify** [comment]
[I'm heading OOO tomorrow](https://expensify.slack.com/archives/C01SKUP7QR0/p1684776773511559) for about a week so adding another BZ member while I'm away to help keep the train moving! 🚂  Thanks @jliexpensify!!

**Comment by Julesssss** [comment]
Okay, back on this today. 

**Setup steps for debugging Android notifications**
- Enable debugging
- Replace the production Airship keys with debug
- Create an Android release build from Android studio

**Notes**
Small messages are correctly retrieved, handled, and threaded.

At a certain point (once the message payload exceeds the current ~2kb limit, we strip onyxData from the payload data, and we fall back to the default notification styling (no threading).

<img width="1349" alt="Screenshot 2023-05-25 at 15 02 47" src="https://github.com/Expensify/App/assets/10736861/8a7bebbd-19b1-4857-9e9e-115943f282a8">
 
It is extremely easy to hit the current 2kb limit, a message of just 15 characters triggers the onyx data to be stripped from a group chat message.
 
| Message | payload size (kb) |
| - | - |
| 1 char | 1899 |
| 15 char | 1994 |
| 50 chars | 2242 |
| 500 chars | 4977 |

**Comment by Julesssss** [comment]
**What is causing us to exceed the payload limit?**  

Here's a breakdown of a few example notifications. As a reminder, any fullsize over 2000 is being removed from the notification payload

| Example | payload size (kb) |
| - | - |
| 10 chars | <img width="462" alt="payload-10chars" src="https://github.com/Expensify/App/assets/10736861/87fc9d85-420a-46f4-ab2d-2d7e9b73d4ce"> |
| 10 emoji | <img width="463" alt="payload-10emoji" src="https://github.com/Expensify/App/assets/10736861/0729280f-7f03-4342-9081-8ec8a83c1d4a"> |
| 100 chars | <img width="465" alt="payload-100chars" src="https://github.com/Expensify/App/assets/10736861/51b47c56-0ea2-4232-87e7-d5ed5a60341c"> |
| 500 chars | <img width="463" alt="payload-500chars" src="https://github.com/Expensify/App/assets/10736861/b81d672e-b0af-4ef2-9a59-0e3b362cc387"> |
| 1000 chars | <img width="464" alt="payload-1000chars" src="https://github.com/Expensify/App/assets/10736861/0553c13a-80fd-4061-bb2b-5fad2a174079"> |

- Notice that the reportAction is roughly 40+% of the payload, this is not used in NewDot!
- We are supposed to first [try to remove the reportAction](https://github.com/Expensify/Web-Expensify/blob/9013ced2a6171923fa156fc75577343f34d7878b/lib/MobilePushNotifications.php#L38-L41), but this doesn't appear to be working

**Comment by Julesssss** [comment]
I went a bit further and sized up the individual onyxData fields: 

- For newDot, we can completely drop reportAction for an immediate 40% increase as it is a duplicate of data already included in the onyxUpdate
- Within the Onyx update, `message` and `originalMessage` are the largest fields by far, followed by `person`, `lastMessageText`, and `avatar` 

| Example | payload size (kb) |
| - | - |
| 100 chars | ![payload-full-100](https://github.com/Expensify/App/assets/10736861/3138df8b-33cb-4c84-9cad-ba5680e8188a) |
| 500 chars | ![payload-full-500](https://github.com/Expensify/App/assets/10736861/77480140-0a77-43ef-acd8-800826b748b9) |
| 1000 chars | ![payload-full-1000](https://github.com/Expensify/App/assets/10736861/0efff870-acec-4241-88c2-1c7416801ea9) |

**Comment by Julesssss** [comment]
Here's a WIP Pr for solving [tasks](https://github.com/Expensify/App/issues/14715#issuecomment-1564458090) 1 and 2.

**Comment by Julesssss** [comment]
I didn't have much time for this after the mobile deployer duty today, but I was able to resolve some style/psalm issues and applied some change suggestions form the WIP PR.

I also prepared custom iOS and Android builds so that I can verify the Web-E changes tomorrow.

**Comment by Julesssss** [comment]
Awesome, I'm seeing great results from solutions 1-3 [listed above](https://github.com/Expensify/App/issues/14715#issuecomment-1564458090)...  once we start zipping the payload we'll have an even larger message length limit:


https://github.com/Expensify/App/assets/10736861/fcd42440-aa5d-4c0e-b138-b4cfc26b305c

**Comment by Julesssss** [comment]
Here are the results of our first improvements. Notifications for messages up to roughly 980 chars in length will now persist the message successfully into onyx, up from 25!

| Example | old size (b) | old wasPersisted (x < 2000) | new size (b) | new wasPersisted (x < 3500) | 
| - | - | - | - | - |
| 10 chars | 1869 | ✅ | 1256 |  ✅ |
| 10 emoji | 2639 | ❌ | 1794|  ✅ |
| 100 chars | 2490 | ❌ | 1600|  ✅ |
| 500 chars | 4884 | ❌ | 2803|  ✅ |
| 1000 chars | 7890 | ❌ | 4303 | ❌  |

**Comment by Julesssss** [comment]
The first PR resolved issues https://github.com/Expensify/Expensify/issues/286640 &
https://github.com/Expensify/Expensify/issues/286641

Continuing with improved logging next

**Comment by Julesssss** [comment]
I think we could close this one now. We'll create follow up issue for further improvements, but the major issue has been resolved.

**Comment by alexxxwork** [comment]
> @alexxxwork thanks for your help so far. We're going to take this internal and work with Callstack and Airship to resolve this given that it has gone 4+ weeks, but we'll be compensating you for your help so far

@cead22 It this issue still eligible for compensation?

**Comment by priya-zha** [comment]
@cead22 @dylanexpensify I'm the external reporter for this issue.

**Comment by Julesssss** [comment]
@dylanexpensify could you please pay out the reporting bonus and 'compensation' based on [this comment](https://github.com/Expensify/App/issues/14715#issuecomment-1599655466) please. 

Sorry all, I have been working on similar issues for a while and didn't realise it was initially raised externally.

**Comment by dylanexpensify** [comment]
Upwork job [here](https://www.upwork.com/jobs/~012c06f8fbdbab58ab)!

@priya-zha please apply for reporting!
@alexxxwork please apply for the work you've done for $500!

**Comment by priya-zha** [comment]
@dylanexpensify Proposal submitted. Thanks.

**Comment by alexxxwork** [comment]
@dylanexpensify applied, thanks

**Comment by dylanexpensify** [comment]
@alexxxwork offer sent!

**Comment by Julesssss** [comment]
@dylanexpensify is this paid out? If so can we close it please

**Comment by dylanexpensify** [comment]
Done!
